libs/wingc/src/lsp/hover.rs

```diff
@@ -1,13 +1,189 @@
-use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind};
+use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind, Position};
 
-use crate::lsp::ast_traversal::find_symbol;
+use crate::ast::{Class, Constructor, Expr, FunctionDefinition, Reference, Scope, Stmt, StmtKind, Symbol};
+use crate::diagnostic::WingSpan;
 use crate::lsp::sync::FILES;
+use crate::visit::Visit;
 use crate::wasm_util::WASM_RETURN_ERROR;
 use crate::{
 	ast::ExprKind,
 	wasm_util::{ptr_to_string, string_to_combined_ptr},
 };
 
+pub struct HoverVisitor<'a> {
+	pub position: Position,
+	pub current_scope: Option<&'a Scope>,
+	pub current_expr: Option<&'a Expr>,
+	pub found_symbol: Option<&'a Symbol>,
+}
+
+impl<'a> HoverVisitor<'a> {
+	pub fn new(position: Position) -> Self {
+		Self {
+			position,
+			current_scope: None,
+			current_expr: None,
+			found_symbol: None,
+		}
+	}
+
+	fn is_found(&self) -> bool {
+		self.found_symbol.is_some()
+	}
+
+	fn should_check_span(&self, span: &'a WingSpan) -> bool {
+		span.contains(&self.position)
+	}
+
+	fn with_scope(&mut self, scope: &'a Scope, mut f: impl FnMut(&mut Self)) {
+		let last_scope = self.current_scope;
+		self.current_scope = Some(scope);
+		f(self);
+		if !self.is_found() {
+			self.current_scope = last_scope;
+		}
+	}
+}
+
+impl<'a> Visit<'a> for HoverVisitor<'a> {
+	fn visit_scope(&mut self, node: &'a Scope) {
+		if self.is_found() {
+			return;
+		}
+
+		self.with_scope(node, |v| {
+			for stmt in &node.statements {
+				v.visit_stmt(stmt);
+			}
+		});
+	}
+
+	fn visit_stmt(&mut self, node: &'a Stmt) {
+		if self.is_found() || !self.should_check_span(&node.span) {
+			return;
+		}
+
+		// Handle situations where symbols are actually defined in inner scopes
+		match &node.kind {
+			StmtKind::ForLoop {
+				iterator,
+				iterable,
+				statements,
+			} => {
```

Comment:
Ah, I see why this is needed... I guess it has to do with the fact that this statement is introducing a new symbol, but that symbol isn't part of the parent scope like the visitor expects/wants in the general case. Feels like a modeling issue but I'm not quite sure how to fix it